crates/sargon/src/profile/mfa/security_structures/roles/builder/roles_builder.rs

```diff
@@ -0,0 +1,870 @@
+            .map(|factor_source_id| {
+                let validation_status = self
+                    .validation_for_addition_of_factor_source_to_list(
+                        factor_source_id,
+                        factor_list_kind,
+                    );
+                FactorSourceInRoleBuilderValidationStatus::new(
+                    self.role(),
+                    *factor_source_id,
+                    validation_status,
+                )
+            })
+            .collect()
+    }
+
+    fn validation_for_addition_of_factor_source_to_list(
+        &self,
+        factor_source_id: &FactorSourceID,
+        factor_list_kind: FactorListKind,
+    ) -> RoleBuilderMutateResult {
+        if self.contains_factor_source(factor_source_id) {
+            return RoleBuilderMutateResult::forever_invalid(
+                FactorSourceAlreadyPresent,
+            );
+        }
+        let factor_source_kind = factor_source_id.get_factor_source_kind();
+        self._validation_add(factor_source_kind, factor_list_kind)
+    }
+
+    fn contains_factor_source(
+        &self,
+        factor_source_id: &FactorSourceID,
+    ) -> bool {
+        self.override_contains_factor_source(factor_source_id)
+            || self.threshold_contains_factor_source(factor_source_id)
+    }
+
+    fn contains_factor_source_of_kind(
+        &self,
+        factor_source_kind: FactorSourceKind,
+    ) -> bool {
+        self.override_contains_factor_source_of_kind(factor_source_kind)
+            || self.threshold_contains_factor_source_of_kind(factor_source_kind)
+    }
+
+    /// Lowers the threshold if the deleted factor source is in the threshold list
+    /// and if after removal of `factor_source_id` `self.threshold > self.threshold_factors.len()`
+    ///
+    /// Returns `Ok` if `factor_source_id` was found and deleted. However, does not call `self.validate()`,
+    /// So state might still be invalid, i.e. we return the result of the action of removal, not the
+    /// state validation status.
+    pub(crate) fn remove_factor_source(
+        &mut self,
+        factor_source_id: &FactorSourceID,
+    ) -> RoleBuilderMutateResult {
+        if !self.contains_factor_source(factor_source_id) {
+            return RoleBuilderMutateResult::basic_violation(
+                FactorSourceNotFound,
+            );
+        }
+        let remove = |xs: &mut Vec<FactorSourceID>| {
+            let index = xs
+                    .iter()
+                    .position(|f| f == factor_source_id)
+                    .expect("Called remove of non existing FactorSourceID, this is a programmer error, should have checked if it exists before calling remove.");
+            xs.remove(index);
+        };
+
+        if self.override_contains_factor_source(factor_source_id) {
+            remove(self.mut_override_factors())
+        }
+        if self.threshold_contains_factor_source(factor_source_id) {
+            remove(self.mut_threshold_factors());
```

Comment:
Not sure this is what we want, users will remove a given factor source from a given factor list kind. Removing the factor source from threshold should not auto remove it from override.

Reply:
A factor can not be present in both lists. So both `if`s can never be both true.

Reply:
I can change to `else if` if you prefer?

Reply:
yeah right, `else if` would be better IMO